src/core/ext/resolver/dns/c_ares/grpc_ares_ev_driver_posix.c

```diff
@@ -0,0 +1,187 @@
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include <grpc/support/port_platform.h>
+#ifndef GRPC_NATIVE_ADDRESS_RESOLVE
+#ifdef GPR_POSIX_SOCKET
+
+#include "src/core/ext/resolver/dns/c_ares/grpc_ares_ev_driver.h"
+
+#include "src/core/lib/iomgr/ev_posix.h"
+#include "src/core/lib/iomgr/sockaddr.h"
+
+#include <ares.h>
+#include <grpc/support/alloc.h>
+#include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <grpc/support/time.h>
+#include <grpc/support/useful.h>
+#include "src/core/lib/iomgr/iomgr_internal.h"
+#include "src/core/lib/iomgr/sockaddr_utils.h"
+#include "src/core/lib/iomgr/unix_sockets_posix.h"
+#include "src/core/lib/support/block_annotate.h"
+#include "src/core/lib/support/string.h"
+
+typedef struct fd_pair {
+  grpc_fd *grpc_fd;
+  int fd;
+  struct fd_pair *next;
+} fd_pair;
+
+struct grpc_ares_ev_driver {
+  bool closing;
+  ares_socket_t socks[ARES_GETSOCK_MAXNUM];
+  int bitmask;
+  grpc_closure driver_closure;
+  grpc_pollset_set *pollset_set;
+  ares_channel channel;
+  fd_pair *fds;
+};
+
+grpc_error *grpc_ares_ev_driver_create(grpc_ares_ev_driver **ev_driver,
+                                       grpc_pollset_set *pollset_set) {
+  int status;
+  *ev_driver = gpr_malloc(sizeof(grpc_ares_ev_driver));
+  status = ares_init(&(*ev_driver)->channel);
+  if (status != ARES_SUCCESS) {
+    gpr_free(*ev_driver);
+    return GRPC_ERROR_CREATE("Failed to init ares channel");
+  }
+  (*ev_driver)->pollset_set = pollset_set;
+  (*ev_driver)->fds = NULL;
+  (*ev_driver)->closing = false;
+  return GRPC_ERROR_NONE;
+}
+
+void grpc_ares_ev_driver_destroy(grpc_ares_ev_driver *ev_driver) {
+  ev_driver->closing = true;
+}
+
+static fd_pair *get_fd(fd_pair **head, int fd) {
+  fd_pair dummy_head;
+  fd_pair *node;
+  fd_pair *ret;
+
+  dummy_head.next = *head;
+  node = &dummy_head;
+  while (node->next != NULL) {
```

Comment:
n is typically 1 - 3 in our tests. Its maximum possible value is ARES_GETSOCK_MAXNUM (which is 16 in current c-ares version).